contracts/Aavegotchi/facets/ERC721BuyOrderFacet.sol

```diff
@@ -0,0 +1,215 @@
+        address indexed buyer,
+        address seller,
+        address erc721TokenAddress,
+        uint256 erc721TokenId,
+        uint256 priceInWei,
+        uint256 time
+    );
+
+    function getERC721BuyOrder(uint256 _buyOrderId) external view returns (ERC721BuyOrder memory buyOrder_) {
+        buyOrder_ = s.erc721BuyOrders[_buyOrderId];
+        require(buyOrder_.timeCreated != 0, "ERC721BuyOrder: ERC721 buyOrder does not exist");
+    }
+
+    function getERC721BuyOrderIdsByTokenId(
+        address _erc721TokenAddress,
+        uint256 _erc721TokenId
+    ) external view returns (uint256[] memory buyOrderIds_) {
+        buyOrderIds_ = s.erc721TokenToBuyOrderIds[_erc721TokenAddress][_erc721TokenId];
+    }
+
+    function getERC721BuyOrdersByTokenId(
+        address _erc721TokenAddress,
+        uint256 _erc721TokenId
+    ) external view returns (ERC721BuyOrder[] memory buyOrders_) {
+        uint256[] memory buyOrderIds = s.erc721TokenToBuyOrderIds[_erc721TokenAddress][_erc721TokenId];
+        uint256 length = buyOrderIds.length;
+        buyOrders_ = new ERC721BuyOrder[](length);
+        for (uint256 i; i < length; i++) {
+            buyOrders_[i] = s.erc721BuyOrders[buyOrderIds[i]];
+        }
+    }
+
+    function placeERC721BuyOrder(
+        address _erc721TokenAddress,
+        uint256 _erc721TokenId,
+        uint256 _priceInWei,
+        uint256 _duration,
+        bool[] calldata _validationOptions // 0: BRS, 1: GHST, 2: skill points
+    ) external {
+        require(_priceInWei >= 1e18, "ERC721BuyOrder: price should be 1 GHST or larger");
+
+        address sender = LibMeta.msgSender();
+        uint256 ghstBalance = IERC20(s.ghstContract).balanceOf(sender);
+        require(ghstBalance >= _priceInWei, "ERC721BuyOrder: Not enough GHST!");
+
+        uint256 category = LibAavegotchi.getERC721Category(_erc721TokenAddress, _erc721TokenId);
+        require(category != LibAavegotchi.STATUS_VRF_PENDING, "ERC721BuyOrder: Cannot buy a portal that is pending VRF");
+        require(sender != s.aavegotchis[_erc721TokenId].owner, "ERC721BuyOrder: Owner can't be buyer");
+        if (category == LibAavegotchi.STATUS_AAVEGOTCHI) {
+            require(_validationOptions.length == 3, "ERC721BuyOrder: Not enough validation options for aavegotchi");
+        }
+
+        uint256 oldBuyOrderId = s.buyerToBuyOrderId[_erc721TokenAddress][_erc721TokenId][sender];
+        if (oldBuyOrderId != 0) {
+            ERC721BuyOrder memory erc721BuyOrder = s.erc721BuyOrders[oldBuyOrderId];
+            require(erc721BuyOrder.timeCreated != 0, "ERC721BuyOrder: ERC721 buyOrder does not exist");
+            require((erc721BuyOrder.cancelled == false) && (erc721BuyOrder.timePurchased == 0), "ERC721BuyOrder: Already processed");
+            if ((erc721BuyOrder.duration == 0) || (erc721BuyOrder.timeCreated + erc721BuyOrder.duration >= block.timestamp)) {
+                LibBuyOrder.cancelERC721BuyOrder(oldBuyOrderId);
```

Comment:
Didn't we discuss updating the amount of the old order instead of cancelling and replacing? It'll be pretty gas expensive to do it this way @orionstardust